File: AIClient_Python/src/aiclient/AI.py
'''
Created on Dec 21, 2014

@author: samuel
'''

from enum import Enum
from aiclient.Singleton import Singleton
from event.QueueController import QueueController
from world.World import World
from mathUtils.Vector2 import Vector2

class AIStatus(Enum):
    '''
    enumememememememe
    '''
    INIT, LOWER_RIGHT, LOWER_LEFT, UPPER_RIGHT, UPPER_LEFT = range(5)

class AI(object):
    '''
    Represents the A.I. that need to be implemented
    
    Attributes:
      aiStatus	the status of the ai (Can be deleted, usefull only for the exemple)
      world	the world that contain all information that the A.I. need
    '''
    aiStatus = AIStatus.INIT
    queueController = Singleton(QueueController)
    world = Singleton(World)
    '''
    Represent la world
    '''

    position1 = Vector2(0,0)
    position2 = Vector2(7,0)
    position3 = Vector2(0,7)
    position4 = Vector2(7,7)
    '''
    Represent la position 4
    '''
    
    '''
    classdocs
    '''
    def tick(self):
        '''
        Function call every 30 ms, this is the starting point for the A.I.
        '''
        if self.aiStatus == AIStatus.INIT:
            self.init()
        elif self.aiStatus == AIStatus.LOWER_RIGHT:
            self.lowerRight()
        elif self.aiStatus == AIStatus.LOWER_LEFT:
            self.lowerLeft()
        elif self.aiStatus == AIStatus.UPPER_RIGHT:
            self.upperRight()
        elif self.aiStatus == AIStatus.UPPER_LEFT:
            self.upperLeft()
    
    def init(self):
        character1 = self.world.getMyTeam().getFirstCharacter()
        character2 = self.world.getMyTeam().getSecondCharacter()
        
        character1.goTo(self.position2)
        character2.goTo(self.position3)
        
        character1.shootRight()
        character2.shootLeft()
        
        self.aiStatus = AIStatus.LOWER_RIGHT
    
    def lowerRight(self):
        character1 = self.world.getMyTeam().getFirstCharacter()
        character2 = self.world.getMyTeam().getSecondCharacter()
        
        if character1.position == self.position2:
            character1.goTo(self.position1)
            character2.goTo(self.position4)

            character1.dropMine()
            character2.dropMine()
            
            character1.shootLeft()
            character2.shootRight()
            
            self.aiStatus = AIStatus.LOWER_LEFT
    
    def lowerLeft(self):
        character1 = self.world.getMyTeam().getFirstCharacter()
        character2 = self.world.getMyTeam().getSecondCharacter()
        
        if character1.position == self.position1:
            character1.goTo(self.position4)
            character2.goTo(self.position1)
            
            character1.shootUp()
            character2.shootDown()
            
            character1.dropMine()
            character2.dropMine()
            
            self.aiStatus = AIStatus.UPPER_RIGHT
    
    def upperRight(self):
        character1 = self.world.getMyTeam().getFirstCharacter()
        character2 = self.world.getMyTeam().getSecondCharacter()
        
        if character1.position == self.position4:
            character1.goTo(self.position3)
            character2.goTo(self.position2)

            character1.shootDown()
            character2.shootUp()

            character1.dropMine()
            character2.dropMine()
            
            self.aiStatus = AIStatus.UPPER_LEFT
    
    def upperLeft(self):
        character1 = self.world.getMyTeam().getFirstCharacter()
        character2 = self.world.getMyTeam().getSecondCharacter()
        
        if character1.position == self.position3:
            character1.goTo(self.position2)
            character2.goTo(self.position3)
            
            character1.shootRight()
            character2.shootLeft()
            
            character1.dropMine()
            character2.dropMine()
            
            self.aiStatus = AIStatus.LOWER_RIGHT

File: AIClient_JAVA/src/aiclient/AI.java
/* ------------------------------------------------------------------------------
 ** _________   _________      ________    _____      _____  ___________ _________
 ** \_   ___ \ /   _____/     /  _____/   /  _  \    /     \ \_   _____//   _____/
 ** /    \  \/ \_____  \     /   \  ___  /  /_\  \  /  \ /  \ |    __)_ \_____  \ 
 ** \     \____/        \    \    \_\  \/    |    \/    Y    \|        \/        \
 **  \______  /_______  /     \______  /\____|__  /\____|__  /_______  /_______  /
 **         \/        \/             \/         \/         \/        \/        \/ 
 **
 ** AI.java
 ** The AI loop
 **
 ** Author: Samuel-Ricardo Carriere
 ** ------------------------------------------------------------------------------*/

package aiclient;

import java.awt.Point;

import world.Character;
import world.Missile;
import world.World;
import event.QueueController;

public class AI {
	public enum AIStatus {
		INIT, LOWER_RIGHT, LOWER_LEFT, UPPER_RIGHT, UPPER_LEFT
	}

	AIStatus status = AIStatus.INIT;
	QueueController queueController = QueueController.getInstance();
	World world = World.getInstance();

	Point position1 = new Point(0, 0);
	Point position2 = new Point(7, 0);
	Point position3 = new Point(0, 7);
	Point position4 = new Point(7, 7);

	public void tick() {
		if (status == AIStatus.INIT) {
			init();
		} else if (status == AIStatus.LOWER_RIGHT) {
			lowerRight();
		} else if (status == AIStatus.LOWER_LEFT) {
			lowerLeft();
		} else if (status == AIStatus.UPPER_RIGHT) {
			upperRight();
		} else if (status == AIStatus.UPPER_LEFT) {
			upperLeft();
		}
	}

	public void init() {
		System.out.println("Init");
		Character king = world.getMyTeam().getCharacter(0);
		Character soldier = world.getMyTeam().getCharacter(1);

		king.move(position2);
		soldier.move(position3);

		king.shootMissile(Missile.Direction.RIGHT);
		soldier.shootMissile(Missile.Direction.LEFT);

		status = AIStatus.LOWER_RIGHT;
	}

	public void lowerRight() {
		Character king = world.getMyTeam().getCharacter(0);
		Character soldier = world.getMyTeam().getCharacter(1);

		if (king.getPosition().equals(position2)) {
			System.out.println("LR");
			king.move(position1);
			soldier.move(position4);

			king.dropMine();
			soldier.dropMine();

			king.shootMissile(Missile.Direction.LEFT);
			soldier.shootMissile(Missile.Direction.RIGHT);

			status = AIStatus.LOWER_LEFT;
		}
	}

	public void lowerLeft() {
		Character king = world.getMyTeam().getCharacter(0);
		Character soldier = world.getMyTeam().getCharacter(1);

		if (king.getPosition().equals(position1)) {
			System.out.println("LL");
			king.move(position4);
			soldier.move(position1);

			king.shootMissile(Missile.Direction.UP);
			soldier.shootMissile(Missile.Direction.DOWN);

			king.dropMine();
			soldier.dropMine();

			status = AIStatus.UPPER_RIGHT;
		}
	}

	public void upperRight() {
		Character king = world.getMyTeam().getCharacter(0);
		Character soldier = world.getMyTeam().getCharacter(1);

		if (king.getPosition().equals(position4)) {
			System.out.println("UR");
			king.move(position3);
			soldier.move(position2);

			king.shootMissile(Missile.Direction.DOWN);
			soldier.shootMissile(Missile.Direction.UP);

			king.dropMine();
			soldier.dropMine();

			status = AIStatus.UPPER_LEFT;
		}
	}

	public void upperLeft() {
		Character king = world.getMyTeam().getCharacter(0);
		Character soldier = world.getMyTeam().getCharacter(1);

		if (king.getPosition().equals(position3)) {
			System.out.println("UL");
			king.move(position2);
			soldier.move(position3);

			king.dropMine();
			soldier.dropMine();

			king.shootMissile(Missile.Direction.RIGHT);
			soldier.shootMissile(Missile.Direction.LEFT);

			status = AIStatus.LOWER_RIGHT;
		}
	}
}
